python/tank/util/metrics.py

```diff
@@ -0,0 +1,391 @@
+
+        """
+
+        super(_MetricsDispatchWorkerThread, self).__init__()
+
+        self._tk = tk
+        self._engine = engine
+
+        # Make this thread a daemon. This means the process won't wait for this
+        # thread to complete before exiting. In most cases, proper engine
+        # shutdown should halt the worker correctly. In cases where an engine
+        # is improperly shut down, this will prevent the process from hanging.
+        self.daemon = True
+
+        # makes possible to halt the thread
+        self._halt_event = Event()
+
+    def run(self):
+        """Runs a loop to dispatch metrics that have been logged."""
+
+        # run until halted
+        while not self._halt_event.isSet():
+
+            # get the next available metric and dispatch it
+            try:
+                metrics = MetricsQueueSingleton().get_metrics(
+                    self.DISPATCH_BATCH_SIZE)
+                if metrics:
+                    self._dispatch(metrics)
+            except Exception, e:
+                self._engine.log_error("Error dispatching metrics: %s" % (e,))
+            finally:
+                # wait, checking for halt event before more processing
+                self._halt_event.wait(self.DISPATCH_INTERVAL)
+
+    def halt(self):
+        """Indiate that the worker thread should halt as soon as possible."""
+        self._halt_event.set()
+
+    def _dispatch(self, metrics):
+        """Dispatch the supplied metric to the sg api registration endpoint.
+
+        :param Metric metrics: The Toolkit metric to dispatch.
+
+        """
+
+        # get this thread's sg connection via tk api
+        sg_connection = self._tk.shotgun
+
+        # handle proxy setup by pulling the proxy details from the main
+        # shotgun connection
+        if sg_connection.config.proxy_handler:
+            opener = urllib2.build_opener(sg_connection.config.proxy_handler)
+            urllib2.install_opener(opener)
+
+        # build the full endpoint url with the shotgun site url
+        url = "%s/%s" % (sg_connection.base_url, self.API_ENDPOINT)
+
+        # construct the payload with the auth args and metrics data
+        payload = {
+            "auth_args": {
+                "session_token": sg_connection.get_session_token()
+            },
+            "metrics": [m.data for m in metrics]
+        }
+        payload_json = json.dumps(payload)
+
+        header = {'Content-Type': 'application/json'}
+        try:
+            request = urllib2.Request(url, payload_json, header)
+            response = urllib2.urlopen(request)
+        except urllib2.HTTPError, e:
+            # fire and forget, so if there's an error, ignore it.
+            pass
+        else:
+            for metric in metrics:
+                self._engine.log_debug("Logged metric: %s" % (metric,))
+
+        # execute the log_metrics core hook
+        self._tk.execute_core_hook(
+            platform_constants.TANK_LOG_METRICS_HOOK_NAME,
```

Comment:
Normally I'd say yes we should do it in the main thread, but given the nature of this hook... I'm not positive.  Feels like anybody implementing this hook would want it to follow our guidelines for it running in the background and we can make up for it with good docs... thoughts?